### DISCUSSION TITLE
Strange error message when using the simulator server?

### DISCUSSION THREAD

**Discussion by heymoonman** [pymodbus-dev/pymodbus#1501] [Help] [answered]
Could be worth an issue but I'm not sure. 

When using the ModbusSimulatorServer I appear to get an error internal to the package regarding a missing log file. This file is supposedly meant to exist inside the PyModbus package itself from what I can tell, which means I'm stuck for being able to fix this locally. 

Error:
![image](https://user-images.githubusercontent.com/34864012/232822065-f140ee99-0bdb-4a4d-afae-8ca69f352290.png)

Could someone please confirm that I'm running the simulator correctly and whether this issue is actually of the package? Thanks.

Here is the code I'm running with a pretty standard setup in the setup json file:

```
import asyncio

from pymodbus.server import ModbusSimulatorServer

async def start_simulator_server():
    server = ModbusSimulatorServer(
        modbus_server='shift_server',
        modbus_device='shift_modbus',
        log_file="",
        json_file='simulator_config.json'
    )
    await server.run_forever()

if __name__ == '__main__':
    asyncio.run(start_simulator_server())
```

I have tried different settings for the kwarg log_file too and found no success. Kind of at a loss, think it may be a package issue where the open() needs to create the log file if it does not exist.

**Comment by janiversen** [comment]
Seems it cannot open your config file. The handling is changed on dev, so it might be worth while to test with dev.

**Comment by heymoonman** [comment]
I'm not sure it is to do with the config file. The open() statement it's referencing occurs in the __init__ function for the ModbusSimulatorServer class:

```
for entry in self.generator_html:  # pylint: disable=consider-using-dict-items
  html_file = os.path.join(self.web_path, "generator", entry)
  with open(html_file, encoding="utf-8") as handle:
      self.generator_html[entry][0] = handle.read()
```

I ran a test on the dev branch of the repo and you're right it didn't throw the same error, though the handling I show above where the error occurs seems to actually be the same so I'm a little baffled.

As a relevant side, is this class actually not meant to be used? The example 'server_simulator.py' instantiates simulator contexts but actually uses the StartAsyncTcpServer function to start it. Is this the recommended way? I've been struggling to follow the examples/docs and make a simulator for a modbus so I can test against it in a test suite.

**Reply by janiversen** [reply]
that open statement open the generator html code, stored in pymodbus/server/simulator/web/generator. These files are the dynamic html pages and are used.

the example shows how to use the datastore simulator, which is without http.

The web front end documentation https://pymodbus.readthedocs.io/en/latest/source/library/simulator/web.html is pretty clear on which class to use.

**Comment by janiversen** [comment, marked as answer]
and just to make it easy, we have included a main ready to run in pymodbus/server/simulator/main.py

which can be started from the command line as pymodbus.simulator

so we did not provide extra examples.
